### ISSUE TITLE
[Feature Request]: Delete any downloaded video after X days

### ISSUE THREAD

**Issue by MarcoBuster** [tubearchivist/tubearchivist#337]
### Already implemented?

- [X] I have read through the [wiki](https://github.com/tubearchivist/tubearchivist/wiki).
- [X] I understand the [scope](https://github.com/tubearchivist/tubearchivist/wiki/FAQ) of this project and am aware of the [known limitations](https://github.com/tubearchivist/tubearchivist#known-limitations) and my idea is not already on the [roadmap](https://github.com/tubearchivist/tubearchivist#roadmap).

### Your Feature Request

After some time, the media library can become very big quickly. 
I'm aware that the 'Delete watched videos after X days' option already exists, but I'm subscribed to many channels and I don't necessarily watch all the videos they publish and manually marking ALL of them as "watched" is a pain.

It think it would be useful adding an option like "Delete _any_ downloaded video after X days" to preserve disk space in the long run.

### Your help is needed!

- [ ] Yes I can help with this feature request!

**Comment by bbilly1** [comment]
So obvious lazy solution here would be not to download channels that you don't want to keep. :-) 

Besides that, you also have the *mark channel as watched* button, this will delete all videos of that channel with the *delete watched videos after x days* functionality, at least after fixing #309... (do you want to help with that?)

But I get it, not everybody has endless storage space. So if you want to implement it, please go ahead. You might have already seen it, `date_downloaded` (epoch timestamp) is already getting stored as part of the video metadata, so that should be straight forward to implement. Ideally also on a per channel basis?

Feel free to reach on on Discord if you need help.

**Comment by thisisawesome1994** [comment]
It might also be a possibility to keep only the x amount of newest videos, for example, if you download the 50 latest videos per channel (setting default), when downloading more videos later, the older videos get deleted (older then the latest 50). Thats what I would suggest. Sorry for the somewhat duplicate.

**Comment by thisisawesome1994** [comment]
> ### Already implemented?
> 
> - [X] I have read through the [wiki](https://github.com/tubearchivist/tubearchivist/wiki).
> - [X] I understand the [scope](https://github.com/tubearchivist/tubearchivist/wiki/FAQ) of this project and am aware of the [known limitations](https://github.com/tubearchivist/tubearchivist#known-limitations) and my idea is not already on the [roadmap](https://github.com/tubearchivist/tubearchivist#roadmap).
> 
> ### Your Feature Request
> 
> After some time, the media library can become very big quickly. 
> I'm aware that the 'Delete watched videos after X days' option already exists, but I'm subscribed to many channels and I don't necessarily watch all the videos they publish and manually marking ALL of them as "watched" is a pain.
> 
> It think it would be useful adding an option like "Delete _any_ downloaded video after X days" to preserve disk space in the long run.
> 
> ### Your help is needed!
> 
> - [X] Yes I can help with this feature request!

Are you implementing this? What is the current state of progress regarding this matter?

**Comment by MarcoBuster** [comment]
@thisisawesome1994 No, I'm sorry but I'm not currently implementing this.

**Comment by thisisawesome1994** [comment]
> @thisisawesome1994 No, I'm sorry but I'm not currently implementing this. 

Thats too bad. I cannot code, but I can test it if you would give me instructions. Do you know of a way to make it work. At the same time I hope bbilly1 will implement this rather sooner then later. Would there be a quick fix I can do myself?

**Comment by thisisawesome1994** [comment]
> So obvious lazy solution here would be not to download channels that you don't want to keep. :-) 
> 
> Besides that, you also have the *mark channel as watched* button, this will delete all videos of that channel with the *delete watched videos after x days* functionality, at least after fixing #309... (do you want to help with that?)
> 
> But I get it, not everybody has endless storage space. So if you want to implement it, please go ahead. You might have already seen it, `date_downloaded` (epoch timestamp) is already getting stored as part of the video metadata, so that should be straight forward to implement. Ideally also on a per channel basis?
> 
> Feel free to reach out on Discord if you need help.

Can you implement this? Or at least put it in the goals of the project?

**Comment by BrianCArnold** [comment]
Welp, I actually need this, so I'll probably implement it soon-ish.

The use case came up with a tech youtuber that uploaded a video that got a lot of backlash about a bad take (on the 4060 Ti 8GB) and then deleted the video. I wanted to see what the original video was, what they got wrong, etc. I'd like to be able to make sure that if they delete a video in last $NUM days, that as long as it got archived, I could watch it in that time period, but I don't want to necessarily keep all of their videos indefinitely. 

This seems to me to be best implemented as a checkbox on "Keep for $NUM days" field on both the channel and app wide that by default only deletes watched videos, but if checked, deletes all videos.

Additionally another checkbox that can change behavior slightly, the ability to keep videos that are deleted/unlisted on youtube, even if they would otherwise be deleted.

Basically, in pseudocode:

```
if (channel.keepForDays != null || app.keepForDays != null)
foreach (var video in channel.videos)
{
  if (video.Age > (channel.keepForDays ?? app.keepForDays))
  {
    if (video.Watched || (channel.DeleteUnWatched ?? app.DeleteUnWatched))
    {
      if (video.Status == "Active" || !(channel.KeepSourceDeleted ?? app.KeepSourceDeleted))
      {
        video.Delete();
      }
    }
  }
}
```

**Comment by bbilly1** [comment]
I'd say keep it simple. the distinction between active and not active videos might lead to unexpected results, as that depends on the Metadata refresh schedule, that refreshes after 90 days by default. 

if you implement an additional numeric field, delete after X days, you can reuse a lot of the functionality already there from the delete watched videos. plus you'll be able to combine it, for example delete watched after 3 days and any video after 30 days...

**Comment by bbilly1** [comment]
this has been sitting here for a long time, nobody wanted to take this on. As this project is an archival project first, that should also be the focus of future enhancements. 

Additionally, with the improvement of the API, you can implement this functionality with your own implementation too.
